assembly/buffer/index.ts

```diff
@@ -16,19 +26,89 @@ export class Buffer extends Uint8Array {
   @unsafe static allocUnsafe(size: i32): Buffer {
     // range must be valid
     if (<usize>size > BLOCK_MAXSIZE) throw new RangeError(E_INVALIDLENGTH);
-    let buffer = __alloc(size, idof<ArrayBuffer>());
-    let result = __alloc(offsetof<Buffer>(), idof<Buffer>());
+    return assembleBuffer(__alloc(size, idof<ArrayBuffer>()), 0, size);
+  }
+
+  public static fromArrayBuffer(buffer: ArrayBuffer, byteOffset: i32 = 0, length: i32 = -1): Buffer {
+    length = select(buffer.byteLength, length, length < 0);
+    if (i32(byteOffset < 0) | i32(byteOffset > buffer.byteLength - length)) throw new RangeError(E_INDEXOUTOFRANGE);
+    if (length == 0) return new Buffer(0);
+
+    return assembleBuffer(changetype<usize>(buffer), <usize>byteOffset, <u32>length);
+  }
+
+  public static fromString(value: string, encoding: string = "utf8"): Buffer {
+    let buffer: ArrayBuffer;
+    if (encoding == "utf8" || encoding == "utf-8") {
+      buffer = String.UTF8.encode(value);
+    } else if (encoding == "utf16le") {
+      buffer = String.UTF16.encode(value);
+    } else if (encoding == "hex") {
+      buffer = Buffer.HEX.encode(value);
+    } else if (encoding == "ascii") {
+      buffer = Buffer.ASCII.encode(value);
+    } else {
+      throw new TypeError("Invalid string encoding.");
+    }
+
+    // assemble the buffer
+    return assembleBuffer(changetype<usize>(buffer), 0, buffer.byteLength);
+  }
+
+  public static fromArray<T extends ArrayBufferView>(value: T, offset: i32 = 0, length: i32 = -1): Buffer {
```

Comment:
@dcodeIO what is going to happen to this `extends` clause? Are we going to detach `Array` from `ArrayBufferView`?